File: clients/python-sdk/routing_dispatch_sdk/dispatch_client.py
"""Dispatch client for assigning routes and managing dispatches"""

from typing import List, Dict, Any, Optional
from .base_client import BaseClient


class DispatchClient(BaseClient):
    """Client for dispatch and route assignment operations"""

    def assign_routes(
        self,
        route_assignments: List[Dict[str, str]],
        auto_notify: bool = True,
    ) -> Dict[str, Any]:
        """
        Assign multiple routes to drivers

        Args:
            route_assignments: List of {routeId, driverId} assignments
            auto_notify: Whether to automatically notify drivers

        Returns:
            Dispatch assignment results

        Example:
            >>> client = DispatchClient(api_key="your-key")
            >>> result = client.assign_routes([
            ...     {"routeId": "route-1", "driverId": "driver-1"},
            ...     {"routeId": "route-2", "driverId": "driver-2"},
            ... ])
            >>> print(f"Assigned {len(result['dispatches'])} routes")
        """
        payload = {
            "assignments": route_assignments,
            "autoNotify": auto_notify,
        }

        return self.post("dispatches/assign", data=payload)

    def create_dispatch(
        self,
        route_id: str,
        driver_id: str,
        vehicle_id: str,
        scheduled_start: str,
        notes: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create a new dispatch assignment

        Args:
            route_id: Route ID
            driver_id: Driver ID
            vehicle_id: Vehicle ID
            scheduled_start: ISO 8601 datetime string
            notes: Optional dispatch notes

        Returns:
            Created dispatch data
        """
        payload = {
            "routeId": route_id,
            "driverId": driver_id,
            "vehicleId": vehicle_id,
            "scheduledStart": scheduled_start,
        }

        if notes:
            payload["notes"] = notes

        return self.post("dispatches", data=payload)

    def get_dispatch(self, dispatch_id: str) -> Dict[str, Any]:
        """
        Get dispatch details by ID

        Args:
            dispatch_id: Dispatch ID

        Returns:
            Dispatch details
        """
        return self.get(f"dispatches/{dispatch_id}")

    def list_dispatches(
        self,
        driver_id: Optional[str] = None,
        status: Optional[str] = None,
        date: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        """
        List dispatches with optional filters

        Args:
            driver_id: Filter by driver ID
            status: Filter by status (pending, in_progress, completed, cancelled)
            date: Filter by date (YYYY-MM-DD)
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            List of dispatches
        """
        params = {"limit": limit, "offset": offset}

        if driver_id:
            params["driverId"] = driver_id
        if status:
            params["status"] = status
        if date:
            params["date"] = date

        response = self.get("dispatches", params=params)
        return response.get("dispatches", [])

    def update_dispatch_status(
        self,
        dispatch_id: str,
        status: str,
        notes: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Update dispatch status

        Args:
            dispatch_id: Dispatch ID
            status: New status (in_progress, completed, cancelled)
            notes: Optional status update notes

        Returns:
            Updated dispatch data
        """
        payload = {"status": status}

        if notes:
            payload["notes"] = notes

        return self.put(f"dispatches/{dispatch_id}/status", data=payload)

    def cancel_dispatch(self, dispatch_id: str, reason: Optional[str] = None) -> Dict[str, Any]:
        """
        Cancel a dispatch

        Args:
            dispatch_id: Dispatch ID
            reason: Optional cancellation reason

        Returns:
            Canceled dispatch data
        """
        payload = {}
        if reason:
            payload["reason"] = reason

        return self.post(f"dispatches/{dispatch_id}/cancel", data=payload)

    def get_driver_schedule(
        self,
        driver_id: str,
        start_date: str,
        end_date: str,
    ) -> List[Dict[str, Any]]:
        """
        Get driver's schedule for a date range

        Args:
            driver_id: Driver ID
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)

        Returns:
            List of scheduled dispatches
        """
        params = {
            "driverId": driver_id,
            "startDate": start_date,
            "endDate": end_date,
        }

        response = self.get("dispatches/schedule", params=params)
        return response.get("schedule", [])

File: clients/node-sdk/src/DispatchClient.ts
/**
 * Dispatch client for assigning routes and managing dispatches
 */

import { BaseClient } from './BaseClient';
import { Dispatch, RouteAssignment, SDKConfig } from './types';

export interface CreateDispatchOptions {
  routeId: string;
  driverId: string;
  vehicleId: string;
  scheduledStart: string;
  notes?: string;
}

export interface ListDispatchesOptions {
  driverId?: string;
  status?: 'pending' | 'in_progress' | 'completed' | 'cancelled';
  date?: string;
  limit?: number;
  offset?: number;
}

export interface UpdateDispatchStatusOptions {
  status: 'in_progress' | 'completed' | 'cancelled';
  notes?: string;
}

export class DispatchClient extends BaseClient {
  constructor(config: SDKConfig) {
    super(config);
  }

  /**
   * Assign multiple routes to drivers
   *
   * @example
   * ```typescript
   * const result = await dispatch.assignRoutes([
   *   { routeId: 'route-1', driverId: 'driver-1' },
   *   { routeId: 'route-2', driverId: 'driver-2' },
   * ]);
   * console.log(`Assigned ${result.dispatches.length} routes`);
   * ```
   */
  async assignRoutes(
    assignments: RouteAssignment[],
    autoNotify: boolean = true
  ): Promise<{ dispatches: Dispatch[] }> {
    const payload = {
      assignments,
      autoNotify,
    };

    return this.post<{ dispatches: Dispatch[] }>('dispatches/assign', payload);
  }

  /**
   * Create a new dispatch assignment
   */
  async createDispatch(options: CreateDispatchOptions): Promise<Dispatch> {
    const payload = {
      routeId: options.routeId,
      driverId: options.driverId,
      vehicleId: options.vehicleId,
      scheduledStart: options.scheduledStart,
      ...(options.notes && { notes: options.notes }),
    };

    return this.post<Dispatch>('dispatches', payload);
  }

  /**
   * Get dispatch details by ID
   */
  async getDispatch(dispatchId: string): Promise<Dispatch> {
    return this.get<Dispatch>(`dispatches/${dispatchId}`);
  }

  /**
   * List dispatches with optional filters
   */
  async listDispatches(options: ListDispatchesOptions = {}): Promise<Dispatch[]> {
    const params = {
      limit: options.limit ?? 50,
      offset: options.offset ?? 0,
      ...(options.driverId && { driverId: options.driverId }),
      ...(options.status && { status: options.status }),
      ...(options.date && { date: options.date }),
    };

    const response = await this.get<{ dispatches: Dispatch[] }>(
      'dispatches',
      params
    );
    return response.dispatches;
  }

  /**
   * Update dispatch status
   */
  async updateDispatchStatus(
    dispatchId: string,
    options: UpdateDispatchStatusOptions
  ): Promise<Dispatch> {
    const payload = {
      status: options.status,
      ...(options.notes && { notes: options.notes }),
    };

    return this.put<Dispatch>(`dispatches/${dispatchId}/status`, payload);
  }

  /**
   * Cancel a dispatch
   */
  async cancelDispatch(
    dispatchId: string,
    reason?: string
  ): Promise<Dispatch> {
    const payload = reason ? { reason } : {};
    return this.post<Dispatch>(`dispatches/${dispatchId}/cancel`, payload);
  }

  /**
   * Get driver's schedule for a date range
   */
  async getDriverSchedule(
    driverId: string,
    startDate: string,
    endDate: string
  ): Promise<Dispatch[]> {
    const params = {
      driverId,
      startDate,
      endDate,
    };

    const response = await this.get<{ schedule: Dispatch[] }>(
      'dispatches/schedule',
      params
    );
    return response.schedule;
  }
}
